crates/data_components/src/github.rs

```diff
@@ -0,0 +1,275 @@
+
+impl GithubRestClient {
+    #[must_use]
+    pub fn new(token: &str) -> Self {
+        let client = reqwest::Client::new();
+        GithubRestClient {
+            client,
+            token: token.into(),
+        }
+    }
+
+    pub async fn fetch_files(
+        &self,
+        owner: &str,
+        repo: &str,
+        tree_sha: &str,
+        limit: Option<usize>,
+        schema: SchemaRef,
+    ) -> Result<Vec<RecordBatch>> {
+        let git_tree = self
+            .fetch_git_tree(owner, repo, tree_sha)
+            .await
+            .context(GithubApiSnafu)?;
+        let mut tree: Vec<GitTreeNode> = git_tree
+            .tree
+            .into_iter()
+            .filter(|node| node.node_type == "blob")
+            .collect();
+
+        if let Some(limit) = limit {
+            tree.truncate(limit);
+        }
+
+        let mut name_builder = StringBuilder::new();
+        let mut path_builder = StringBuilder::new();
+        let mut size_builder = UInt64Builder::new();
+        let mut sha_builder = StringBuilder::new();
+        let mut mode_builder = StringBuilder::new();
+        let mut url_builder = StringBuilder::new();
+        for node in tree {
+            name_builder.append_value(extract_name_from_path(&node.path).unwrap_or_default());
+            path_builder.append_value(&node.path);
+            size_builder.append_value(node.size.unwrap_or(0));
+            sha_builder.append_value(&node.sha);
+            mode_builder.append_value(&node.mode);
+            url_builder.append_value(&node.url);
+        }
+
+        let record_batch = RecordBatch::try_new(
+            Arc::clone(&schema),
+            vec![
+                Arc::new(name_builder.finish()),
+                Arc::new(path_builder.finish()),
+                Arc::new(size_builder.finish()),
+                Arc::new(sha_builder.finish()),
+                Arc::new(mode_builder.finish()),
+                Arc::new(url_builder.finish()),
+            ],
+        )
+        .context(UnableToConstructRecordBatchSnafu)?;
+
+        Ok(vec![record_batch])
+    }
+
+    async fn fetch_git_tree(
+        &self,
+        owner: &str,
+        repo: &str,
+        tree_sha: &str,
+    ) -> Result<GitTree, Box<dyn std::error::Error + Send + Sync>> {
+        let endpoint = format!(
+            "https://api.github.com/repos/{owner}/{repo}/git/trees/{tree_sha}?recursive=true"
+        );
+
+        let mut headers = HeaderMap::new();
+        headers.insert(USER_AGENT, HeaderValue::from_static(SPICE_USER_AGENT));
+        headers.insert(
+            ACCEPT,
+            HeaderValue::from_static("application/vnd.github.v3+json"),
+        );
+
+        if let Ok(header) = HeaderValue::from_str(&format!("token {}", self.token)) {
+            headers.insert(AUTHORIZATION, header);
+        }
+
+        tracing::debug!("fetch_git_tree: endpoint: {}", endpoint);
+
+        let response = self.client.get(&endpoint).headers(headers).send().await?;
```

Comment:
This could be easily cached (like SQL result _cache) with cache key (owner, repo, tree_sha). It could also cache subtrees (like `node.sha`, i think?), so consecutive searches for subcomponents get cache hit. 

Github throttles pretty quickly.

Reply:
I like it 👍 Will add to the TODO list